plugins.py

```diff
@@ -0,0 +1,83 @@
+import time
+import datetime
+import deltachat
+
+
+class ReceivePlugin:
+    name = "test account"
+
+    @staticmethod
+    @deltachat.account_hookimpl
+    def ac_incoming_message(message):
+        received = time.time()
+        message.create_chat()
+        if not message.chat.can_send():
+            # if it's a device message or mailing list, we don't need to look at it.
+            return
+        sender = message.get_sender_contact().addr
+        receiver = message.account.get_self_contact().addr
+        msgcontent = parse_msg(message.text)
+        firsttravel = msgcontent.get("testduration")
+        secondtravel = received - msgcontent.get("begin")
+        iam = msgcontent.get("sender")
+        if message.chat.is_group():
+            if iam == "spider":
+                print("%s: joined group chat %s after %.1f seconds" % (receiver, message.chat.get_name(), secondtravel))
+                message.chat.send_text("Sender: %s\nBegin: %s" % (receiver, str(time.time())))
+                return  # for now it's enough to just accept the group chat
+            else:
+                return  # when messages by others arrived is checked later by the main thread
+        print("%s: test message took %.1f seconds to %s and %.1f seconds back." %
+              (receiver, firsttravel, sender, secondtravel))
+        message.account.output.submit_1on1_result(receiver, firsttravel, secondtravel)
+        message.account.shutdown()
+
+
+class EchoPlugin:
+    name = "spider"
+
+    @staticmethod
+    @deltachat.account_hookimpl
+    def ac_incoming_message(message):
+        received = time.time()
+        message.create_chat()
+        addr = message.get_sender_contact().addr
+        if message.is_system_message():
+            message.chat.send_text("echoing system message from {}:\n{}".format(addr, message))
+        elif message.chat.is_group():
+            return  # can safely ignore group messages. spider only creates it
+        else:
+            sent = parse_msg(message.text).get("begin")
+            testduration = received - sent
+            msginfo = message.get_message_info()
+            begin = time.time()
+            message.chat.send_text("TestDuration: %f\nBegin: %f\n%s" % (testduration, begin, msginfo))
+
+
+def parse_msg(text: str) -> dict:
+    """Parse data out of a message.
+
+    :param text: a string containing the message info of a deltachat.message.
+    :return: a dictionary with different values parse from the message info.
+    """
```

Comment:
Originally, this function parsed message info strings. Then I started putting all kinds of message.text strings in there. All of them have a classic format (`key: value`) which makes them quite easy to parse.

I heard that parsers are difficult to write safely; hope I didn't make any dangerous security mistakes.

Reply:
well, this script is not run by many people and in a rather controlled way.  We don't need to be concerned with security much here.